src/header/header_def.rs

```diff
@@ -0,0 +1,446 @@
+                    if id < 1 {
+                        return Err(RTPError::RFC8285TwoByteHeaderIDRange(id));
+                    }
+                    if payload.len() > 255 {
+                        return Err(RTPError::RFC8285TwoByteHeaderSize(id));
+                    }
+                }
+                _ => {
+                    if id != 0 {
+                        return Err(RTPError::RFC3550HeaderIDRange(id));
+                    }
+                }
+            };
+
+            // Update existing if it exists else add new extension
+            for extension in &mut self.extensions {
+                if extension.id == id {
+                    extension.payload.clear();
+                    extension.payload.extend_from_slice(payload);
+                    return Ok(());
+                }
+            }
+            self.extensions.push(Extension {
+                id,
+                payload: payload.to_owned(),
+            });
+            return Ok(());
+        }
+
+        // No existing header extensions
+        self.extension = true;
+
+        let len = payload.len();
+        if len <= 16 {
+            self.extension_profile = super::ExtensionProfile::OneByte as u16
+        } else if len > 16 && len < 256 {
+            self.extension_profile = super::ExtensionProfile::TwoByte as u16
+        }
+
+        self.extensions.push(Extension {
+            id,
+            payload: payload.to_owned(),
+        });
+
+        Ok(())
+    }
+
+    /// Returns an RTP header extension
+    pub fn get_extension(&self, id: u8) -> Option<&[u8]> {
+        if !self.extension {
+            return None;
+        }
+
+        for extension in &self.extensions {
+            if extension.id == id {
+                return Some(&extension.payload);
+            }
+        }
+        None
+    }
+
+    // GetExtensionIDs returns an extension id array
+    pub fn get_extension_ids(&self) -> Vec<u8> {
+        if !self.extension {
+            return vec![];
+        }
+
+        if self.extensions.is_empty() {
+            return vec![];
+        }
+
+        let mut ids = vec![0u8; self.extensions.len()];
+
+        for id in ids.iter_mut().take(self.extensions.len()) {
+            *id = self.extensions[0].id
+        }
+
+        ids
+    }
+
+    /// Removes an RTP Header extension
+    pub fn del_extension(&mut self, id: u8) -> Result<(), RTPError> {
+        if !self.extension {
+            return Err(RTPError::HeaderExtensionNotEnabled);
+        }
+        for index in 0..self.extensions.len() {
+            if self.extensions[index].id == id {
+                self.extensions.remove(index);
+                return Ok(());
+            }
+        }
+
+        Err(RTPError::HeaderExtensionNotFound)
+    }
+
+    /// Unmarshal parses the passed byte slice and stores the result in the Header this method is called upon
+    pub fn unmarshal(&mut self, raw_packet: &mut [u8]) -> Result<usize, RTPError> {
```

Comment:
it is preferred that buf is read-only for unmarshal function.

Reply:
Fixed, thanks.